sound/soc/sof/ipc4-topology.c

```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
 	}
 }
 
+static const struct sof_ipc4_audio_format *
+sof_ipc4_get_widget_pin_format(struct snd_sof_widget *swidget, bool pin_type,
+			       int pin_id)
+{
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct sof_ipc4_base_module_cfg_ext *base_ext = NULL;
+	const struct sof_ipc4_pin_format *pin_fmt;
+	struct sof_ipc4_control_data *control_data;
+	const struct snd_kcontrol_new *kc;
+	struct snd_sof_control *scontrol;
+	struct soc_bytes_ext *sbe;
+	int pin_start, pin_end;
+	int i;
+
+	if (swidget->init_blob_style == SOF_INIT_BLOB_BASE_WITH_MOD_EXT) {
+		struct sof_ipc4_base_module_cfg *base_config = swidget->private;
+
+		if (pin_type == SOF_PIN_TYPE_SOURCE && swidget->payload_with_output_fmt) {
```

Comment:
when will this be true? If I see the call below, pin_type is always SOF_PIN_TYPE_SINK?

Reply:
in updating pipeline params (process module prepare ops), we need to get the output pin format, and set the output format to pipeline param, the pin_type will be SOF_PIN_TYPE_SOURCE there.